web_src/js/features/comp/EditorUpload.ts

```diff
@@ -99,8 +99,12 @@ async function handleUploadFiles(editor, dropzoneEl, files, e) {
     const {width, dppx} = await imageInfo(file);
     const placeholder = `[${name}](uploading ...)`;
 
-    editor.insertPlaceholder(placeholder);
+    // Placeholders should be inserted on uploading the thumnail.
+    // If not, the check upload process may fail
+    const handleInsertPlaceholder = () => editor.insertPlaceholder(placeholder); 
+    dropzoneEl.dropzone.on('thumbnail', handleInsertPlaceholder);
     await uploadFile(dropzoneEl, file); // the "file" will get its "uuid" during the upload
+    dropzoneEl.dropzone.off('thumbnail', handleInsertPlaceholder); 
```

Comment:
It still doesn't seem right.

1. `await` might fail and throw an exception, then the handler is left there.
2. If there are more than 1 files are being uploaded (in progress), then it will trigger more than one `handleInsertPlaceholder` calls, duplicate calls will insert duplicate handlers?

Reply:
I corrected the first point, but I don't know how to solve the second problem. If you think of a way, can you suggest a solution?

Reply:
There is one more bad case for: if a uploading fails, then there will be no onContentChanged callback, then the buttons stay "disabled".

----

I think the UI states should be managed by some well-defined event&handlers, but not by a ad-hoc "thumbnail" with onContentChanged callback.